core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java

```diff
@@ -360,21 +360,36 @@ public Sep38QuoteResponse postQuote(JwtToken token, Sep38PostQuoteRequest reques
 
     // Calculate amounts: sellAmount = buyAmount * totalPrice
     BigDecimal bTotalPrice = decimal(rate.getTotalPrice());
-    BigDecimal bSellAmount, bBuyAmount;
-    if (request.getSellAmount() != null) {
-      bSellAmount = decimal(request.getSellAmount());
-      bBuyAmount = bSellAmount.divide(bTotalPrice, buyAsset.getDecimals(), RoundingMode.HALF_UP);
-    } else {
-      bBuyAmount = decimal(request.getBuyAmount());
+    BigDecimal bSellAmount =
+        request.getSellAmount() != null ? decimal(request.getSellAmount()) : null;
+    BigDecimal bBuyAmount = request.getBuyAmount() != null ? decimal(request.getBuyAmount()) : null;
+
+    // Use the rate's buy and sell amounts if they were returned
+    if (rate.getSellAmount() != null) {
+      bSellAmount = decimal(rate.getSellAmount());
+    }
+    if (rate.getBuyAmount() != null) {
+      bBuyAmount = decimal(rate.getBuyAmount());
+    }
+
+    // This should not happen because we previously checked at least one was not null
+    if (bBuyAmount == null && bSellAmount == null) {
+      throw new ServerErrorException("Unable to calculate buy and sell amounts, both were null");
+    }
+
+    if (bSellAmount == null) {
       bSellAmount = bBuyAmount.multiply(bTotalPrice);
+    } else if (bBuyAmount == null) {
+      bBuyAmount = bSellAmount.divide(bTotalPrice, buyAsset.getDecimals(), RoundingMode.HALF_UP);
```

Comment:
I see we're using `HALF_UP` here but we're using `FLOOR` when we [calculate `totalPrice` in 2.0](https://github.com/stellar/java-stellar-anchor-sdk/blob/ba1df433f66ae1a6fd58d32c350b125af3346008/core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java#L488). We're also using a hard-coded precision.

In a 2.X release we should probably introduce a set of configuration options that allow the anchor platform to calculate and validate amounts instead.

Reply:
@JakeUrban we have a task for that
https://stellarorg.atlassian.net/browse/ANCHOR-280

Reply:
We are using FLOOR because one of your partners (currently on v1) uses this mode to calculate the total price, with precision 10.
So, in should not break them, as it will be a default config.
However, new partners may want to have a different setting, so it should be configurable in the asset config